PortForward: Tunnel SPDY through WebSockets

seans3 opened the pull request:
* Implements PortForward over WebSockets by tunneling the current SPDY implementation through WebSockets.
* Creates a tunneling proxy within the apiserver to tunnel SPDY data from the client to the API Server through WebSockets. The proxy is protected by a feature flag (`PortForwardWebsockets`) and it is only executed for the 
 connection upgrade request with a tunneling protocol (example: `SPDY/3.1+portforward.k8s.io` for tunneling SPDY through WebSockets) subprotocol header.
* On the client, implements tunneling by injecting a WebSockets connection (as a `TunnelingConnection`) into the SPDY connection during construction.
* Another feature flag exists for `kubectl` to create a tunneling SPDY connection: `KUBECTL_PORT_FORWARD_WEBSOCKETS`.

/kind feature

```release-note
- kubectl port-forward over websockets (tunneling SPDY) can be enabled using an `Alpha` feature flag environment variable: KUBECTL_PORT_FORWARD_WEBSOCKETS=true. The API Server being communicated to must *also* have an `Alpha` feature flag enabled: PortForwardWebsockets.
```
- [KEP Transition from SPDY to WebSockets - 4006](https://github.com/kubernetes/enhancements/issues/4006)

seans3 committed:
portforward: tunnel spdy through websockets

seans3 commented:
/sig cli
/sig apimachinery

seans3 commented:
/cc @liggitt

liggitt commented on pkg/registry/core/pod/rest/subresources.go:
```diff
@@ -242,7 +242,13 @@ func (r *PortForwardREST) Connect(ctx context.Context, name string, opts runtime
+	if utilfeature.DefaultFeatureGate.Enabled(features.PortForwardWebsockets) {
+		maxBytesPerSec := capabilities.Get().PerConnectionBandwidthLimitBytesPerSec
+		tunnelingHandler := translator.NewTunnelingHandler(handler, maxBytesPerSec)
+		handler = translator.NewTranslatingHandler(handler, tunnelingHandler, wsstream.IsWebSocketRequestWithPortForwardProtocol)
```
nit: if tunnelingHandler is delegating all data flow serially to handler, I *think* we can just rely on its maxBytesPerSec enforcement

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go:
```diff
@@ -137,6 +137,12 @@ func (f *defaultPortForwarder) ForwardPorts(method string, url *url.URL, opts Po
 	}
 	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, method, url)
+	if cmdutil.PortForwardWebsockets.IsEnabled() {
+		dialer, err = portforward.NewTunnelingDialer(url, opts.Config, opts.Ports)
```
tunneling dialer probably doesn't need opts.Ports

if we fail to create/upgrade the websocket request, we'll need to fall back to the standard spdy dialer, right? that probably means we want a fallback dialer, something like:

```
spdyDialer := ...
if cmdutil.PortForwardWebsockets.IsEnabled() {
  websocketDialer := ...
  dialer = fallbackDialer(websocketDialer, spdyDialer)
} else {
  dialer = spdyDialer
}
```

seans3 replied:
Added a `fallbackDialer` and unit tests. Removed `opts.Ports`. Done.

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,122 @@
+	}
+	defer conn.Close() //nolint:errcheck
+	klog.Infof("websocket connection created: %s", conn.Subprotocol())
+	h.tunnelingConnection = portforward.NewTunnelingConnection(conn, h.MaxBytesPerSec)
```
let's make TunnelingHandler#ServeHTTP non-mutating to TunnelingHandler, keep tunnelingConnection a local var

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+	// 	w = flowrate.NewWriter(w, s.maxBytesPerSec)
+	// }
+	// Next, write the passed data in "p".
+	n, err := w.Write(p)
```
can n be different than len(p)? do we need to loop until we've completely written p?

liggitt replied:
answering my own question, no, it can't

> Write must return a non-nil error if it returns n < len(p)

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+	}
+	defer func() {
+		if w != nil {
+			w.Close() //nolint:errcheck
```
can simplify this with a named return and conditionally setting the return error

func (c *TunnelingConnection) Write(p []byte) (n int, err error) {
...
defer func(){
  // close, which flushes the message
  closeErr := w.Close()
  if closeErr != nil && err == nil {
    // if closing/flushing errored and we weren't already returning an error, return the close error
    err = closeErr
  }
}()
}

seans3 replied:
Fixed. Please have a look.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+func (c *TunnelingConnection) SetDeadline(t time.Time) error {
+	err := c.SetReadDeadline(t)
```
probably call both even if one errors, and return an error if either errored

does this do the right thing with the zero value?

> A zero value for t means I/O operations will not time out.

seans3 replied:
Fixed.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+func (c *TunnelingConnection) RemoteAddr() net.Addr {
+	return nil
```
do we know if spdy makes use of this?

liggitt replied:
actually, looks like we can call `c.conn.RemoteAddr()`

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+func (c *TunnelingConnection) Close() error {
+	klog.Infoln("tunneling connection Close()...")
+	// Signal other endpoint that websocket connection is closing.
+	c.conn.WriteControl(gwebsocket.CloseMessage, []byte{}, time.Now().Add(writeDeadline)) //nolint:errcheck
+	return c.conn.Close()
```
should we do anything to make sure we only call this once?

seans3 replied:
Added `sync.Once` to ensure we only call this close code (for each endpoint) only once.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+	return n, err
+}
+
+func (c *TunnelingConnection) Close() error {
```
should this return an error if WriteControl errors and c.conn.Close() does not?

liggitt replied:
still outstanding

aojea replied:
what was the decision, it does not fail closing a closed connection?

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+func (c *TunnelingConnection) LocalAddr() net.Addr {
+	return nil
```
can return `c.conn.LocalAddr()`

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,165 @@
+	closeChan         chan bool
+	maxBytesPerSec    int64
+	inProgressMessage io.Reader
+}
```
might not need closeChan and maxBytesPerSec in this type

seans3 replied:
Removed both.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_dialer.go:
```diff
@@ -0,0 +1,99 @@
+		klog.V(4).Infof("Remote Port: %d", remotePort)
+		query.Set("ports", strconv.Itoa(remotePort))
+		req.URL.RawQuery = query.Encode()
+	}
```
I don't *think* we need to encode the ports in the URL here since we're tunneling the spdy protocol which sends the ports on stream establishment

seans3 replied:
Removed

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_dialer.go:
```diff
@@ -0,0 +1,99 @@
+		req.URL.RawQuery = query.Encode()
+	}
+	// Hard-code the v2 portforward protocol for the websocket dialer for now.
+	websocketProtocols := []string{"v2.portforward.k8s.io"}
```
the protocols passed in are spdy protocols... this dialer is going to tunnel those over websockets, so I think we could construct the websocket protocol name generically by prefixing `spdy+`, something like this:

```
var spdyTunnelingWebsocketProtocols []string
for _, p := range protocols {
  spdyTunnelingWebsocketProtocols = append(spdyTunnelingWebsocketProtocols, "SPDY/3.1+"+p)
}
conn, err := websocket.Negotiate(d.transport, d.holder, req, websocketProtocols...)
...
```

then whatever the server sends back as the accepted websocket subprotocol, we trim off the `"SPDY/3.1+"` prefix for, and then can return the chosen spdy protocol in a way that makes sense to the caller
```
selectedSpdyProtocol := strings.TrimPrefix(conn.Subprotocol(), "SPDY/3.1+")
...
return spdyConn, selectedSpdyProtocol, err
```

that lets this client tunneler be completely agnostic about whatever SPDY/3.1 stuff is happening over the tunnel

seans3 replied:
We're adding the SPDY protocols (and other request headers) as we call the `UpgradeAwareProxy` passing in the updated `http.Request`. The `UpgradeAwareProxy` is doing the SPDY upgrade using these modified headers. We don't need to do anything special with these WebSocket upgrade headers.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_dialer.go:
```diff
@@ -0,0 +1,99 @@
+// NewTunnelingDialer creates and returns the tunnelingDialer structure which implemements the "httpstream.Dialer"
+// interface. The dialer can upgrade a websocket request, creating a websocket connection. This function
+// returns an error if one occurs.
+func NewTunnelingDialer(url *url.URL, config *restclient.Config, ports []string) (httpstream.Dialer, error) {
```
nit: name this to make it clear it is tunneling SPDY over websockets, e.g. NewSPDYOverWebsocketDialer

seans3 replied:
Fixed.

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,122 @@
+			return true // Accepting all requests
+		},
+		Subprotocols: []string{
+			constants.PortForwardV2Name,
```
can define this as the server-side version of https://github.com/kubernetes/kubernetes/pull/123413/files#r1497897438, e.g. `"SPDY/3.1+"+constants.PortForwardV1Name`

seans3 replied:
I don't think we need to do anything that combines the SPDY/WebSockets headers. This should be just the websocket v2 portforward protocol as sent by the tunneling dialer. After the websocket connection is created in the client (using standard websocket headers, including v2) it is wrapped in the `TunnelingConnection` which implements `net.Conn`. This is what is passed into the creation of the SPDY client, transparently tunneling all SPDY connection calls.

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,122 @@
+
+func (h *TunnelingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	klog.Infoln("TunnelingHandler ServeHTTP")
+	clone := utilnet.CloneRequest(req)
```
might need to nil out or set the body of the cloned request to a Noop reader... we don't expect the body of the request to be read at all by the kubelet, only the hijacked connection

seans3 replied:
Fixed.

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,122 @@
+
+func (h *TunnelingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	klog.Infoln("TunnelingHandler ServeHTTP")
+	clone := utilnet.CloneRequest(req)
```
spdy sends POST, not GET, right? do we need to change that in the clone?

seans3 replied:
Fixed.

seans3 commented:
/triage accepted
/priority important-soon

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_dialer.go:
```diff
@@ -0,0 +1,87 @@
+// containing a WebSockets connection (which implements "net.Conn"). Also
+// returns the protocol negotiated (currently v2 portforward), or an error.
+// The passed "protocols" are ignored.
+func (d *tunnelingDialer) Dial(protocols ...string) (httpstream.Connection, string, error) {
```
protocols passed in are currently `[portforward.k8s.io]`, hard-coded at https://github.com/kubernetes/kubernetes/blob/54bcbc3c750a4cf9f87d10b7a69b64e250d4495e/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L194

the protocol we return *has* to be one of the ones passed in

there's nothing specific about the portforward protocol in this tunneling connection, which is why I suggested passing all the protocols to the server with a SPDY/3.1+ prefix and letting the server decide what we should speak, then stripping that prefix off and returning the selected one to the caller

seans3 replied:
We're using `v2.portforward.k8s.io` to select the `TunnelingHandler` instead of the legacy `UpgradeAwareHandler` 

https://github.com/kubernetes/kubernetes/blob/cb05a861a9a6c180f1fb8f04e853155ded3d5f29/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go#L113

We'll need to take that into account.

seans3 replied:
I updated the tunneling protocol to use the `SPDY/3.1+` prefix approach. Please let me know what you think.

liggitt replied:
looks good, thanks

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/portforward/constants.go:
```diff
@@ -0,0 +1,23 @@
+
+const (
+	PortForwardV1Name   = "portforward.k8s.io"
+	SpdyTunnelingPrefix = "SPDY/3.1+"
+	TunnelPortForwardV1 = SpdyTunnelingPrefix + PortForwardV1Name
```
nit, make it clear this is for use with websockets:

```suggestion
	WebsocketSPDYTunnelingPrefix = "SPDY/3.1+"
	WebsocketSPDYTunnelingPortForwardV1 = WebsocketSPDYTunnelingPrefix + PortForwardV1
```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -370,21 +372,20 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 
 	if backendHTTPResponse.StatusCode != http.StatusSwitchingProtocols {
 		// If the backend did not upgrade the request, echo the response from the backend to the client and return, closing the connection.
-		klog.V(6).Infof("Proxy upgrade error, status code %d", backendHTTPResponse.StatusCode)
```
restore this?

seans3 replied:
Fixed.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -400,7 +401,6 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 
 	// Forward raw response bytes back to client.
 	if len(rawResponse) > 0 {
-		klog.V(6).Infof("Writing %d bytes to hijacked connection", len(rawResponse))
```
restore this?

seans3 replied:
Fixed.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/fallback_dialer.go:
```diff
@@ -0,0 +1,58 @@
+// if one occurs.
+func (f *fallbackDialer) Dial(protocols ...string) (httpstream.Connection, string, error) {
+	conn, version, err := f.primary.Dial(protocols...)
+	klog.V(4).Infof("primary dialer err: %v", err)
+	if f.shouldFallback(err) {
```
```suggestion
	if err != nil && f.shouldFallback(err) {
		klog.V(4).Infof("fallback to secondary dialer for primary dialer err: %v", err)
```

seans3 replied:
Fixed.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,160 @@
+		klog.V(8).Infof("%s: tunneling connection read in progress message...", c.name)
+		i, err := c.inProgressMessage.Read(p)
+		switch {
+		case err == nil:
```
I realized this could drop data in the case where the websocket conn returns `i > 0` AND `err == EOF` (which is allowed and even encouraged by https://pkg.go.dev/io#Reader):

> An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.
> ...
> Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.

I think this should switch to:

```go
if i == 0 && err == io.EOF {
	c.inProgressMessage = nil
} else {
	klog.V(8).Infof("%s: read %d bytes, error=%v, bytes=% X", c.name, i, err, p[:i])
	return i, err
}
```

seans3 replied:
Fixed. Tested using the stress test: 20k requests (=4k in 5 different terminals).

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,160 @@
+func (c *TunnelingConnection) Write(p []byte) (n int, err error) {
+	klog.V(7).Infof("%s: write: %d bytes, bytes=% X", c.name, len(p), p)
+	defer klog.V(7).Infof("%s: tunneling connection write...complete", c.name)
+	if c.conn == nil {
+		return 0, fmt.Errorf("write on closed tunneling connection")
+	}
```
We're not doing anything to nil out the conn when Close() is called, so I don't think we'll currently ever get here? We're also not doing nil checks on `conn` in Read(). I think we should just drop this nil check

seans3 replied:
Removed unused `conn == nil` check.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,160 @@
+	if c.conn == nil {
+		return 0, fmt.Errorf("write on closed tunneling connection")
+	}
+	err = c.SetWriteDeadline(time.Now().Add(writeDeadline))
```
why are we calling this before every write? that undoes any other calls people made to `SetWriteDeadline`

seans3 replied:
Removed.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,160 @@
+	"k8s.io/klog/v2"
+)
+
+const writeDeadline = 2 * time.Second
```
how was this deadline selected? I'm not sure I'd expect us to hard code a write deadline like this

seans3 replied:
This is the same deadline as in the `RemoteCommand` code. The comments are:
```
// writeDeadline defines the time that a write to the websocket connection                                                                                        
// must complete by, otherwise an i/o timeout occurs. The writeDeadline                                                                                           
// has nothing to do with a response from the other websocket connection                                                                                          
// endpoint; only that the message was successfully processed by the                                                                                              
// local websocket connection. The typical write deadline within the websocket                                                                                    
// library is one second.                                                                                                                                         
const writeDeadline = 2 * time.Second     
```

liggitt replied:
since we're only using this in the close connection now, inline this and shorten it to 1 second

separately, I think we need to revisit the RemoteCommand use of this and make sure we're not making the websocket version more vulnerable to network stalls than the spdy version

seans3 replied:
Inlined and shortened write deadline to 1 second for this close signal.

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,189 @@
+	// Update the http request for an upstream SPDY upgrade.
+	clone.Method = "POST"
+	clone.Body = nil // Remove the request body which is unused.
+	clone.Header.Add(httpstream.HeaderUpgrade, spdy.HeaderSpdy31)
+	clone.Header.Add(httpstream.HeaderProtocolVersion, protocol)
```
prefer `Set` to `Add` (we don't want to combine with any existing values in these headers)

seans3 replied:
Fixed.

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,189 @@
+	}
+
+	h.completedHeaders = true
+	h.parsedStatus = resp.Status
```
should we check the resp.Status?

liggitt replied:
definitely, looking at the headers and status that come back to make sure it is an upgrade and make sure the backend selected a spdy protocol is on the list

liggitt replied:
I actually need to do the latter as part of shifting the websocket upgrade back to the client after the backend has indicated it can upgrade

seans3 replied:
The `UpgradeAware` proxy is also handling/validating the SPDY response. We should never get to the `headerInterceptingConnection` (which only happens after `Hijack`) until we're sure it's a `101 Switching Protocols`. I'm storing the status and headers mostly for testing.

liggitt replied:
I'm pretty sure we have to check the selected protocol is valid, but I'll verify

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,189 @@
+
+	// Write into the headerBuffer, then attempt to parse the bytes
+	// as an http response.
+	n, err := h.headerBuffer.Write(b)
+	if err != nil {
+		return n, err
+	}
+	bufferedReader := bufio.NewReader(h.headerBuffer)
```
I think you can also read directly from the connection
```
	bufferedReader := bufio.NewReader(netConn)
```

aojea replied:
ah, ok, you use the buffer later for other things

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -218,54 +224,161 @@ func (w *tunnelingResponseWriter) WriteHeader(statusCode int) {
-		return n, err
+		return len(b), err
 	}
+	resp.Body.Close()
```
need to verify this is correct and doesn't drain the buffered reader (unit tests should exercise this)

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -218,54 +224,161 @@ func (w *tunnelingResponseWriter) WriteHeader(statusCode int) {
+	// record that we closed so we don't write again or try to initialize
+	u.err = fmt.Errorf("connection closed")
+	// write a response
+	http.Error(u.w, u.err.Error(), http.StatusInternalServerError)
```
double-check status code on this

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -218,54 +224,161 @@ func (w *tunnelingResponseWriter) WriteHeader(statusCode int) {
+	u.lock.RLock()
+	defer u.lock.RUnlock()
+	if u.conn != nil {
+		return u.conn.RemoteAddr()
```
LocalAddr

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -218,54 +224,161 @@ func (w *tunnelingResponseWriter) WriteHeader(statusCode int) {
+	if u.conn != nil {
+		return u.conn.RemoteAddr()
+	}
+	// TODO: can we do better here?
```
return noop impl?

seans3 commented:
Looks like a leap-year bug in the `pkg/printers/internalversion`, which should resolve itself tomorrow.
```
{Failed;  === RUN   TestPrintIPAddressList
    printers_test.go:6616: mismatch:   []v1.TableRow{
          	{Cells: {string("192.168.2.2"), string("myresource.mygroup/mynamespace/myname"), string("10y")}},
          	{
          		Cells: []any{
          			string("2001:db8::2"),
          			string("myresource2.mygroup2/mynamespace2/myname2"),
        - 			string("5y1d"),
        + 			string("5y2d"),
          		},
          		Conditions: nil,
          		Object:     {},
          	},
          }
--- FAIL: TestPrintIPAddressList (0.00s)
;}
```

liggitt commented:
> Looks like a leap-year bug in the `pkg/printers/internalversion`, which should resolve itself tomorrow (since today is a leap day 2/29/2024).

fix in https://github.com/kubernetes/kubernetes/pull/123603

seans3 commented:
/retest

seans3 commented:
/retest

Unrelated error in required test
```
{Failed;Failed;Failed;  === RUN   TestPolicyAdmission/.v1.bindings/create
I0302 01:35:38.881821  109885 policy_source.go:224] refreshing policies
    admission_test_util.go:342: version: v1beta1, phase:validation, converted:true error: no request received
    admission_test_util.go:342: version: v1beta1, phase:validation, converted:false error: no request received
        --- FAIL: TestPolicyAdmission/.v1.bindings/create (0.06s)
;=== RUN   TestPolicyAdmission/.v1.bindings
    --- FAIL: TestPolicyAdmission/.v1.bindings (0.06s)
```

seans3 committed:
adds return context in RunPortForward()

aojea commented on staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go:
```diff
@@ -420,10 +420,14 @@ func (o PortForwardOptions) RunPortForwardContext(ctx context.Context) error {
 	signal.Notify(signals, os.Interrupt)
 	defer signal.Stop(signals)
 
+	returnCtx, returnCtxCancel := context.WithCancel(context.Background())
```
if you inherit the context, then you should just need only `case <-returnCtx.Done():` in the select

dims commented:
/milestone v1.30

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -452,6 +454,18 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 	return true
 }
 
+// loggingReader logs the bytes read from the "delegate" with a "name" prefix.
+type loggingReader struct {
+	name     string
+	delegate io.Reader
+}
+
+func (l *loggingReader) Read(p []byte) (int, error) {
+	n, err := l.delegate.Read(p)
+	klog.V(8).Infof("%s: %d bytes, err=%v, bytes=% X", l.name, n, err, p[:n])
+	return n, err
+}
+
```
I think these served their purpose and we can / should remove them now to avoid logging proxied data at any verbosity level

seans3 replied:
Removed.

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go:
```diff
@@ -0,0 +1,153 @@
+		klog.V(7).Infof("%s: tunneling connection Close()...", c.name)
+		// Signal other endpoint that websocket connection is closing; ignore error.
+		normalCloseMsg := gwebsocket.FormatCloseMessage(gwebsocket.CloseNormalClosure, "")
+		c.conn.WriteControl(gwebsocket.CloseMessage, normalCloseMsg, time.Now().Add(time.Second)) //nolint:errcheck
+		err = c.conn.Close()
```
if either WriteControl or Close error, we should return an error:

```suggestion
		writeControlErr := c.conn.WriteControl(gwebsocket.CloseMessage, normalCloseMsg, time.Now().Add(time.Second))
		closeErr := c.conn.Close()
		if closeErr != nil {
		  err = closeErr
		} else if writeControlErr != nil {
		  err = writeControlErr
		}
```

seans3 replied:
Fixed

liggitt committed:
Fix headerInterceptingConn handling

seans3 committed:
removes extra upgrade aware proxy logging; returns tunneling connection close error

aojea commented on staging/src/k8s.io/client-go/tools/portforward/fallback_dialer_test.go:
```diff
@@ -0,0 +1,53 @@
+
+func TestFallbackDialer(t *testing.T) {
+	protocol := "v6.fake.k8s.io"
+	// If "shouldFallback" is false, then only primary should be dialed.
+	primary := &fakeDialer{dialed: false}
+	primary.negotiatedProtocol = protocol
```
You have to return an error on the primary dial or the  `alwaysFalse` condition check will not be exercised

seans3 replied:
I re-wrote this entire test to be more useful. Please have a look.

aojea commented on staging/src/k8s.io/client-go/tools/portforward/fallback_dialer_test.go:
```diff
@@ -0,0 +1,53 @@
+	assert.True(t, primary.dialed, "no fallback; primary should have dialed")
+	assert.Equal(t, protocol, negotiated, "")
+	assert.False(t, secondary.dialed, "no fallback; secondary should *not* have dialed")
+	assert.Nil(t, err, "error should be nil")
```
if err == nil, the dial succeeded , we want to test that we don't fallback if dial primary fails, right?

seans3 replied:
I re-wrote this entire test to be more useful. Please have a look.

aojea commented on pkg/registry/core/pod/rest/subresources.go:
```diff
@@ -242,7 +242,12 @@ func (r *PortForwardREST) Connect(ctx context.Context, name string, opts runtime
+	handler := newThrottledUpgradeAwareProxyHandler(location, transport, false, true, responder)
+	if utilfeature.DefaultFeatureGate.Enabled(features.PortForwardWebsockets) {
+		tunnelingHandler := translator.NewTunnelingHandler(handler)
+		handler = translator.NewTranslatingHandler(handler, tunnelingHandler, wsstream.IsWebSocketRequestWithTunnelingProtocol)
```
this abstraction is really nice, code reads great

```
grep -r NewTranslatingHandler pkg/
pkg/registry/core/pod/rest/subresources.go:             handler = translator.NewTranslatingHandler(handler, streamtranslator, wsstream.IsWebSocketRequestWithStreamCloseProtocol)
pkg/registry/core/pod/rest/subresources.go:             handler = translator.NewTranslatingHandler(handler, streamtranslator, wsstream.IsWebSocketRequestWithStreamCloseProtocol)
pkg/registry/core/pod/rest/subresources.go:             handler = translator.NewTranslatingHandler(handler, tunnelingHandler, wsstream.IsWebSocketRequestWithTunnelingProtocol)
```

aojea commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -191,11 +191,15 @@ func (pf *PortForwarder) ForwardPorts() error {
 	}
 	defer pf.streamConn.Close()
+	if protocol != PortForwardProtocolV1Name {
+		return fmt.Errorf("unable to negotiate protocol: client supports %q, server returned %q", PortForwardProtocolV1Name, protocol)
```
I assume is safe to strict the validation now

aojea commented on staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go:
```diff
@@ -387,7 +400,14 @@ func (o PortForwardOptions) Validate() error {
 
 // RunPortForward implements all the necessary functionality for port-forward cmd.
 func (o PortForwardOptions) RunPortForward() error {
```
TODO follow up we may indicate to people they should use the new one
```
	// Deprecated: Use RunPortForwardContext instead, which allows 
	// to cancel port forwarding.
```

seans3 replied:
Added `Deprecation` warning to the function comments.

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+	}
+	w.hijacked = true
+	klog.V(6).Infof("Hijack returning websocket tunneling net.Conn")
+	return w.conn, nil, nil
```
should we return the bufio.ReadWriter ?

https://cs.opensource.google/go/go/+/master:src/net/http/server.go;l=315;drc=6f5d77454e31be8af11a7e2bcda36d200fda07c5

aojea replied:
perfect

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+// Hijack returns a delegate "net.Conn".
+// An error is returned if Write(), WriteHeader(), or Hijack() was previously called.
+// The returned bufio.ReadWriter is always nil.
+func (w *tunnelingResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
```
The std lib https://cs.opensource.google/go/go/+/faf882d1d427e8c8a9a1be00d8ddcab81d1e848e abort pending reads, before hijacking, I don't feel we have to, do we ?

liggitt replied:
since in our use we don't do anything with the request, I don't think so

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+}
+
+// Write is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingResponseWriter) Write(p []byte) (int, error) {
```
should we account for this behavior?
> 	// WriteHeader sends an HTTP response header with the provided
	// status code.
	//
	// If WriteHeader is not called explicitly, the first call to Write
	// will trigger an implicit WriteHeader(http.StatusOK).
	// Thus explicit calls to WriteHeader are mainly used to
	// send error codes or 1xx informational responses.

https://go.dev/src/net/http/server.go?h=WriteHeader

aojea replied:
oh, nvm, this is delegated, so the delegated ResponseWriter will do it for us

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+		w: w,
+		conn: &headerInterceptingConn{
+			initializableConn: &tunnelingWebsocketUpgraderConn{
+				w:   w,
```
this `w http.ResponseWriter` is passed around many times is hard to think about it, but it seems meanwhile `w: w` and `conn` here are thread safe we are safe

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+	InitializeWrite(backendResponse *http.Response, backendResponseBytes []byte) error
+}
+
+const maxHeaderBytes = 1 << 20
```
Note for self 
https://github.com/golang/go/blob/e39480f4fb5e7314daef012b58cf05e3c9ceea1a/src/net/http/server.go#L868-L871

```
// DefaultMaxHeaderBytes is the maximum permitted size of the headers
// in an HTTP request.
// This can be overridden by setting [Server.MaxHeaderBytes].
const DefaultMaxHeaderBytes = 1 << 20 // 1 MB
```

aojea commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+
+const maxHeaderBytes = 1 << 20
+
+// token for normal header / body separation (\r\n\r\n, but go tolerates the leading \r being absent)
+var lfCRLF = []byte("\n\r\n")
+
+// token for header / body separation without \r (which go tolerates)
+var lfLF = []byte("\n\n")
```
it took me more than 5 mins to find this https://github.com/golang/go/issues/53186 , did you already know about this behavior?

liggitt replied:
nope, my unit tests flushed it out

seans3 committed:
re-write fallback dialer unit test

seans3 committed:
adds comments to tunnelingResponseWriter

seans3 commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go:
```diff
@@ -0,0 +1,454 @@
+
+	writer := &tunnelingResponseWriter{
+		w: w,
+		conn: &headerInterceptingConn{
```
Done.

aojea commented:
/retest

> There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors ((combined from similar events): 0/161 nodes are available: 161 Insufficient cpu, 20 Insufficient memory. preemption: 0/161 nodes are available: 161 No preemption victims found for incoming pod..)


/lgtm

liggitt commented:
/lgtm
/approve
/retest

please update the release note to describe the alpha feature availability and use

liggitt commented:
/retest

may the node scheduling odds be ever in your favor 🎲

seans3 commented:
/retest

> /lgtm /approve /retest
> 
> please update the release note to describe the alpha feature availability and use

I have added a release note describing the use of the `Alpha` feature flags. Please have a look.

seans3 commented:
/retest

liggitt commented:
Maybe https://github.com/kubernetes/test-infra/pull/32156 will help with test scheduling

BenTheElder commented:
scheduling: https://github.com/kubernetes/test-infra/issues/32157

seans3 commented:
/retest pull-kubernetes-e2e-kind

seans3 commented:
/test pull-kubernetes-e2e-kind

seans3 commented:
/test pull-kubernetes-e2e-kind

aojea commented:
great job
